File: Leetcode/Typescript/150. Evaluate Reverse Polish Notation/solution.ts
// https://leetcode.com/problems/evaluate-reverse-polish-notation/

function evalRPN(tokens: string[]): number {
  const nums: number[] = [];

  for (const t of tokens) {
    if (!isNaN(Number(t))) {
      nums.push(parseInt(t));
    } else {
      const num = nums.pop();

      if (num === undefined) {
        throw new Error("Invalid expression");
      }
      const oper = t;

      switch (oper) {
        case "+":
          nums[nums.length - 1] += num;
          break;
        case "-":
          nums[nums.length - 1] -= num;
          break;
        case "*":
          nums[nums.length - 1] *= num;
          break;
        case "/":
          nums[nums.length - 1] = Math.trunc(nums[nums.length - 1] / num);
          break;
      }
    }
  }

  return nums.pop();
}

File: Leetcode/C++/150. Evaluate Reverse Polish Notation/solution.cpp
// https://leetcode.com/problems/evaluate-reverse-polish-notation/description/

class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        stack<int> st;

        for (string &token : tokens)
        {
            if (token == "+" || token == "-" || token == "*" || token == "/")
            {
                int b = st.top();
                st.pop();
                int a = st.top();
                st.pop();
                int result;

                if (token == "+")
                    result = a + b;
                else if (token == "-")
                    result = a - b;
                else if (token == "*")
                    result = a * b;
                else if (token == "/")
                    result = a / b; // 정수 나눗셈, trunc toward 0

                st.push(result);
            }
            else
            {
                st.push(stoi(token));
            }
        }

        return st.top();
    }
};
